aloha-vw-jni/src/main/scala/com/eharmony/aloha/models/vw/jni/multilabel/VwSparseMultilabelPredictor.scala

```diff
@@ -0,0 +1,182 @@
+import com.eharmony.aloha.io.sources.ModelSource
+import com.eharmony.aloha.models.multilabel.SparseMultiLabelPredictor
+import vowpalWabbit.learner.{VWActionScoresLearner, VWLearners}
+import vowpalWabbit.responses.ActionScores
+
+import scala.collection.{immutable => sci}
+import scala.util.Try
+
+/**
+  *
+  * Created by ryan.deak on 9/8/17.
+  * @param modelSource a specification for the underlying ''Cost Sensitive One Against All''
+  *                    VW model with ''label dependent features''.  VW flag
+  *                    `--csoaa_ldf mc` is expected. For more information, see the
+  *                    [[https://github.com/JohnLangford/vowpal_wabbit/wiki/Cost-Sensitive-One-Against-All-(csoaa)-multi-class-example VW CSOAA wiki page]].
+  *                    Also see the ''Cost-Sensitive Multiclass Classification'' section of
+  *                    Hal Daume's [[https://www.umiacs.umd.edu/%7Ehal/tmp/multiclassVW.html On Multiclass Classification in VW]]
+  *                    page.  This model specification will be materialized in this class.
+  * @param params VW parameters.
+  * @param defaultNs The list of indices into the `features` sequence that does not have
+  *                  an exist in any value of the `namespaces` map.
+  * @param namespaces Mapping from namespace name to indices in the `features` sequence passed
+  *                   to the `apply` method.  There should be no empty namespaces, meaning
+  *                   ''key-value'' pairs appearing in the map should not values that are empty
+  *                   sequences.  '''This is a requirement.'''
+  * @tparam K the label or class type.
+  */
+// TODO: Comment this function.  It requires a lot of assumptions.  Make those known.
+case class VwSparseMultilabelPredictor[K](
+    modelSource: ModelSource,
+
+    // TODO: Should these be removed?  I don't think so but could be, w/o harm, in limited cases.
+    params: String,
+    defaultNs: List[Int],
+    namespaces: List[(String, List[Int])],
+    numLabelsInTrainingSet: Int)
+extends SparseMultiLabelPredictor[K]
+   with Closeable {
+
+  import VwSparseMultilabelPredictor._
+
+  @transient private[this] lazy val paramsAndVwModel =
+    createLearner(modelSource, params, numLabelsInTrainingSet)
+
+  @transient private[this] lazy val updatedParams = paramsAndVwModel._1
+  @transient private[multilabel] lazy val vwModel = paramsAndVwModel._2.get
+
+
+  {
+    val emptyNss = namespaces collect { case (ns, ind) if ind.isEmpty => ns }
+    require(
+      emptyNss.isEmpty,
+      s"There should be no namespaces that are empty.  Found: ${emptyNss mkString ", "}"
+    )
+
+    // Force creation.
+    require(vwModel != null)
+  }
+
+  /**
+    * Get the VW parameters used to invoke the underlying VW model.
+    * @return VW parameters.
+    */
+  def vwParams(): String = updatedParams
+
+  /**
+    * Given the input, form a VW example, and delegate to the underlying ''CSOAA LDF'' VW model.
```

Comment:
Not here, but in general we should say why we MUST use a CSOAA LDF model and not just a CSOAA model when that could be sufficient.  We should comment on how VW doesn't export those types correctly through the JNI and how that's the fault not of the JNI, but of the CSOAA class on the C side.  We should also comment on how we'll add more support that for CSOAA which is much simpler once that work is done within VW.